src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,370 @@
+        validate(elements);
+
+        List<ColumnMetadata> columns = columns();
+        for (int i = 0, m = columns.size(); i < m; i++)
+        {
+            ColumnMetadata column = columns.get(i);
+            ByteBuffer element = elements.get(i);
+            if (element == null)
+                throw invalidRequest("Invalid null value for %s in %s",
+                                     column.name.toCQLString(), columnsExpression);
+            if (element == ByteBufferUtil.UNSET_BYTE_BUFFER)
+                throw invalidRequest("Invalid unset value for %s in %s",
+                                     column.name.toCQLString(), columnsExpression);
+        }
+    }
+
+    @Override
+    public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryOptions options)
+    {
+        if (isOnToken())
+            throw new UnsupportedOperationException();
+
+        switch (columnsExpression.kind())
+        {
+            case SINGLE_COLUMN:
+                List<ByteBuffer> buffers = bindAndGet(options);
+
+                ColumnMetadata column = firstColumn();
+                if (operator == Operator.IN)
+                {
+                    filter.add(column, operator, inValues(column, buffers));
+                }
+                else if (operator == Operator.LIKE)
+                {
+                    LikePattern pattern = LikePattern.parse(buffers.get(0));
+                    // there must be a suitable INDEX for LIKE_XXX expressions
+                    RowFilter.SimpleExpression expression = filter.add(column, pattern.kind().operator(), pattern.value());
+                    indexRegistry.getBestIndexFor(expression)
+                                 .orElseThrow(() -> invalidRequest("%s is only supported on properly indexed columns",
+                                                                   expression));
+                }
+                else
+                {
+                    filter.add(column, operator, buffers.get(0));
+                }
+                break;
+            case MULTI_COLUMN:
+                checkFalse(isSlice(), "Multi-column slice restrictions cannot be used for filtering.");
+
+                if (isEQ())
+                {
+                    List<ByteBuffer> elements = bindAndGetElements(options).get(0);
+
+                    for (int i = 0, m = columns().size(); i < m; i++)
+                    {
+                        ColumnMetadata columnDef = columns().get(i);
+                        filter.add(columnDef, Operator.EQ, elements.get(i));
+                    }
+                }
+                else if (isIN())
+                {
+                    // If the relation is of the type (c) IN ((x),(y),(z)) then it is equivalent to
+                    // c IN (x, y, z) and we can perform filtering
+                    if (columns().size() == 1)
+                    {
+                        List<ByteBuffer> values = bindAndGetElements(options).stream()
```

Comment:
Streams in `IN` queries.

Reply:
It is a tiny corner case that is probably never hit.